shub/settings.py

```diff
@@ -0,0 +1,662 @@
+    # Anyone in this group is a superuser for the app
+    "AUTH_LDAP_SUPERUSER_GROUP_FLAGS": None,  # "cn=superuser,ou=django,ou=groups,dc=example,dc=com"
+    # "cn=sregistry_admin,ou=groups,dc=example,dc=com"
+    # Globus Assocation (globus)
+    # Only required if 'globus' is added to PLUGINS_ENABLED in config.py
+    "SOCIAL_AUTH_GLOBUS_KEY": None,
+    "SOCIAL_AUTH_GLOBUS_USERNAME": None,  # "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx@clients.auth.globus.org"
+    "SOCIAL_AUTH_GLOBUS_SECRET": None,
+    "GLOBUS_ENDPOINT_ID": None,
+    # SAML Authentication (saml)
+    # Only required if 'saml_auth' is added to PLUGINS_ENABLED
+    "AUTH_SAML_IDP": None,  # "stanford"
+    "AUTH_SAML_INSTITUTION": None,  # "Stanford University"
+}
+
+INTEGER_DEFAULTS = {
+    "API_PAGE_SIZE": 10,
+    # Set this to be some size in MB to limit uploads.
+    # Uploads > 2.5GB will not use memory, but the filesystem
+    "DATA_UPLOAD_MAX_MEMORY_SIZE": None,
+    # Limit users to N collections (None is unlimited)
+    "USER_COLLECTION_LIMIT": 2,
+    # The number of collections to show on the /<domain>/collections page
+    "COLLECTIONS_VIEW_PAGE_COUNT": 250,
+    # The maximum number of downloads allowed per container/collection, per week
+    "CONTAINER_WEEKLY_GET_LIMIT": 100,
+    "COLLECTION_WEEKLY_GET_LIMIT": 100,
+    "GOOGLE_BUILD_LIMIT": 100,
+    "GOOGLE_BUILD_TIMEOUT_SECONDS": None,  # None defaults to 10 minutes
+    "GOOGLE_BUILD_EXPIRE_SECONDS": 28800,
+    "MINIO_SIGNED_URL_EXPIRE_MINUTES": 5,
+    # Google Build
+    # To prevent denial of service attacks on Google Cloud Storage, you should set a reasonable limit for the number of active, concurrent builds.
+    # This number should be based on your expected number of users, repositories, and recipes per repository.
+    "GOOGLE_BUILD_LIMIT": None,  # 100
+    # The number of seconds for the build to timeout. If set to None, will be 10 minutes. If
+    # unset, will default to 3 hours. This time should be less than the GOOGLE_BUILD_EXPIRE_SECONDS
+    "GOOGLE_BUILD_TIMEOUT_SECONDS": None,
+    # The number of seconds for the build to expire, meaning it's response is no longer accepted by the server. This must be defined.
+    # Using 28800 would indicate 8 hours (in seconds)
+    "GOOGLE_BUILD_EXPIRE_SECONDS": None,  # 28800
+    # The number of seconds to expire a signed URL given to download a container
+    # from storage. This can be much smaller than 10, as we only need it to endure
+    # for the POST.
+    "CONTAINER_SIGNED_URL_EXPIRE_SECONDS": None,  # 10
+}
+
+# Environment helpers
+
+
+def get_sregistry_envar(key):
+    envar_key = "SREGISTRY_%s" % key
+    return os.environ.get(envar_key)
+
+
+def get_sregistry_envar_list(key):
+    key = "SREGISTRY_%s" % key
+    envar_list = os.environ.get(key) or []
+    if envar_list:
+        envar_list = [x.strip() for x in envar_list.split(",") if x.strip()]
+    return envar_list
+
+
+# Try reading in from secrets first (no issue if not found)
```

Comment:
I would move this to the end of the file so that the load order maps to the conceptual "define and then override" that we're doing:
1. defaults
2. environment
3. settings.yaml
4. secrets.py

The secrets currently take precedence because of this code on line 664:
```python
    # Don't set if the value is empty, or it's been set previously
    if value is None or key in locals() and locals()[key] is not None:
```

I think the order you have is correct and moving the secrets imports to line 648 makes it very clear that the file trumps everything else.

Reply:
Ah ok, so you agree about environment being honored first? I think my point of confusion is that I'm loading in this order:

1. secrets.py
2. environment
3. settings.yaml
4. defaults

and the implementation does that by using that order, but never setting something that is already defined.

Reply:
Agree - we can put secrets at the bottom, although I'll want to double check / think about if any of the previous settings (e.g., putting a value into a data structure) would prevent that.

Reply:
okay I think this will work - with the assumption that secrets.py will work only for values you could previously put there (e.g., the new values for the various API settings that I'm adding newly will not work) and to over-ride entire datastructures (e.g., the database) you need to define the entire thing (and it's not exposed just via attributes of that).